source/hpack/encoder.d

```diff
@@ -0,0 +1,265 @@
+module hpack.encoder;
+
+import hpack.tables;
+import hpack.huffman;
+import hpack.util;
+import hpack.exception;
+
+import vibe.http.status;
+import vibe.http.common;
+
+import std.range;
+import std.bitmanip; // prefix encoding / decoding
+import std.typecons;
+import std.conv;
+import std.exception;
+import std.array;
+
+struct HeaderEncoder(T = HTTP2HeaderTableField[])
+	if(isInputRange!T && is(ElementType!T : HTTP2HeaderTableField) ||
+	   is(T == HTTP2HeaderTableField))
+{
+	private {
+		static if(is(T == HTTP2HeaderTableField)) T[] m_range;
+		else T m_range;
+		IndexingTable m_table;
+		ubyte[] m_encoded;
+	}
+
+	// flag to select huffman encoding / raw encoding
+	bool huffman;
```

Comment:
It could change according to the HTTP/2 response to be sent, that's why it's not private.
AFAIK, there's no fixed way of choosing between Huffman encoding or raw encoding. As an alternative the `huffman` variable could be left private and made available only in the constructor, so that one would need to initialize two encoders, one for huffman and one for raw.

Reply:
Okay, I assumed that you'd instantiate a new encoder per request/response since you'd usually provide a new input range of headers, but that makes sense. It does however provide some potential for misuse due to the asynchronous nature of the input vs. the output ranges (i.e. when exactly changing the field affects the output stream is not necessarily clear).